### PR TITLE
Dashboard configurator

### PR DESCRIPTION
**Pull request by adambraimbridge**
image

@wheresrhys there's a few todos remaining for this, but ready for initial discussion/feedback.

### CODE REVIEW

File: views/partials/chart-configurator.html [hunk 1]
```diff
@@ -2,10 +2,62 @@
...
+			<input name="timeframe" type="radio" class="o-forms-radio o-forms-radio--small"
+				id="this_6_months_{{index}}"
+				value="this_6_months" {{#ifEquals @root.timeframe 'this_6_months'}}checked{{/ifEquals}}>
+			<label class="o-forms-label" for="this_6_months_{{index}}">6 months</label>
+
+			<input name="timeframe" type="radio" class="o-forms-radio o-forms-radio--small"
+				id="this_12_months_{{index}}"
+				value="this_12_months" {{#ifEquals @root.timeframe 'this_12_months'}}checked{{/ifEquals}}>
+			<label class="o-forms-label" for="this_12_months_{{index}}">12 months</label>
+
```
Comment (adambraimbridge) [inline views/partials/chart-configurator.html]: Todo: Tidy up the configurator form (maybe change these radios to a select) (I can't believe I'm suggesting a select GAH)

File: client/components/chart.js [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+}
+
+export function displayError (printerEl, err, kq, meta) {
+	console.log('Error', err, kq, meta);
+	printerEl.classList.remove('chart--loading');
+	printerEl.classList.add('chart-error');
+	printerEl.innerHTML = `<p class="error"><strong>Error: </strong>${err.message || err}</span><p>${meta.name}, ${meta.label}, ${meta.question}: ${meta.query}</p>`;
+}
+
+export function renderChart (printerEl, builtQuery, chart) {
```
Comment (wheresrhys) [inline client/components/chart.js]: I'd much prefer to keep the third variable as `chartConfig` or something similar

File: client/components/chart.js [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+		// avoid race condition when clicking multiple configurator buttons in quick succession
+		if (rendererMap.get(printerEl) !== renderPromise) {
+			return;
+		}
+		displayError(printerEl, err, builtQuery, chart);
+	});
+}
+
+export function buildAndRenderChart (chart, printerEl) {
+	try {
```
Comment (wheresrhys) [inline client/components/chart.js]: I'd much prefer to keep the first variable as `chartConfig` or something similar

File: client/components/configurator.js [hunk 4]
```diff
@@ -64,6 +64,14 @@ function getStateFromQuery () {
 		};
 	}
 	q.filters = [];
+
+	if (q.pagetype) {
+		q.filters.push({
+			prop:'page.location.type',
+			value:q.pagetype
```
Comment (wheresrhys) [inline client/components/configurator.js]: I'm not 100% what I think of this, but in general I think it's only worth creating aliases for existing property names if the property name stored in keen is really unclear what it means. 'page.location.type'  could just be referred to as 'page.location.type' everywhere, which'd make it easier to reuse this code for other properties

File: views/partials/chart-configurator.html [hunk 5]
```diff
@@ -2,10 +2,62 @@
...
-			<input class="o-forms-radio o-forms-radio--small" id="this_48_hours_{{index}}" name="timeframe" type="radio" value="this_48_hours" {{#ifEquals @root.timeframe 'this_48_hours'}}checked{{/ifEquals}}><label class="o-forms-label" for="this_48_hours_{{index}}">2 days</label>
-			<input class="o-forms-radio o-forms-radio--small" id="this_7_days_{{index}}" name="timeframe" type="radio" value="this_7_days" {{#ifEquals @root.timeframe 'this_7_days'}}checked{{/ifEquals}}><label class="o-forms-label" for="this_7_days_{{index}}">7 days</label>
-			<input class="o-forms-radio o-forms-radio--small" id="this_14_days_{{index}}" name="timeframe" type="radio" value="this_14_days" {{#ifEquals @root.timeframe 'this_14_days'}}checked{{/ifEquals}}><label class="o-forms-label" for="this_14_days_{{index}}">14 days</label>
-			<input class="o-forms-radio o-forms-radio--small" id="this_30_days_{{index}}" name="timeframe" type="radio" value="this_30_days" {{#ifEquals @root.timeframe 'this_30_days'}}checked{{/ifEquals}}><label class="o-forms-label" for="this_30_days_{{index}}">30 days</label>
+
+			<input name="timeframe" type="radio" class="o-forms-radio o-forms-radio--small"
```
Comment (wheresrhys) [inline views/partials/chart-configurator.html]: Can we maintain an array of timeframe + interval combinations somewhere, add it to res.locals in middleware and then use it do populate the timeframe pickers. Makes it easier to change later if it turns out people actually want to use.

**Comment by wheresrhys** [comment]
:+1: apart from comments about the 'chart' variable. Other comments are just suggestions

File: client/components/configurator.js (hunk 4 above)
Comment (adambraimbridge) [inline client/components/configurator.js]: Definitely agree. I'm thinking dashboard URLs ought to have query parameters identical to keen explorer, where appropriate. I'll do that in a future commit.

File: client/components/chart.js (hunk 3 above)
Comment (adambraimbridge) [inline client/components/chart.js]: No problemo at all. My only preference is to be consistent with variable names as much as possible. We still have lots of instances of "charts" (e.g. `window.charts`, `chartsMap`) that perhaps should be renamed to "chartConfigs". I've previously used `alias`/`aliases`, `meta`, and thought about calling them something more abstract, like `posit`s — but `chartConfig` works for me.

**Pull request merged**